### PR TITLE
Catch ReplicaSet Deletion Events

### LINKED ISSUE #69376: ReplicaSetController can miss handling the deletion of a ReplicaSet
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:

@ncdc's description hoisted from https://github.com/kubernetes/kubernetes/issues/69376#issuecomment-427453571:

I was about to file this. I spent yesterday triaging it and have determined the root cause. The issue is that the `ReplicaSetController` can "miss" handling the deletion of a `ReplicaSet` if things happen quickly enough. Here's the flow:

1. Client creates rs
1. `ReplicaSetController` sees new rs, starts working on creating pods
1. Client deletes rs
1. `ReplicaSetController`'s `rsInformer` see the rs deletion and calls `rsc.enqueueReplicaSet`, which adds the namespace/name of the rs to the work queue
1. Client recreates rs with the exact same name as before
1. `ReplicaSetController`'s `rsInformer` see the rs addition and calls `rsc.enqueueReplicaSet`, which adds the namespace/name of the rs to the work queue (again)
1. `ReplicaSetController`'s sync handler processes the entry from the queue
1. Because the rs was recreated with the same name, when `syncReplicaSet` calls `rsLister` to get the rs, it's there (it's the 2nd one)

This is a timing issue. The `ReplicaSetController` doesn't check the rs's UID, and if the order of operations is "just right", the controller's sync handler won't "see" the deletion, so it never calls `rsc.expectations.DeleteExpectations(key)` to reflect that the rs was deleted.

/sig apps

Original description follows:

In #69344 I had to re-run the integration test 4 times to get it to work.

Log outputs:

* [attempt 1](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/69344/pull-kubernetes-integration/29684/)
* [attempt 2](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/69344/pull-kubernetes-integration/29685/)
* [attempt 3](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/69344/pull-kubernetes-integration/29688/)
* [attempt 4 (successful)](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/69344/pull-kubernetes-integration/29691/)

### PR DESCRIPTION
**Pull request by runyontr**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #69376

**Special notes for your reviewer**:

When the ReplicaSet controller fetches a fresh version of the ReplicaSet from the server
it validates that the UID of the fresh matches the one in the cache.  If the
UID does not match, a race condition outlined in the issue must have occurred and the
controller should interpret the call to syncReplicaSet as a delete.

Reverts changes from https://github.com/kubernetes/kubernetes/pull/69739 that temporarily fixed tests to not be flaky

* Does this fix need to get rolled into other Controllers?
* If so, should we move these into common packages now?

**Does this PR introduce a user-facing change?**:
None

**Relaese Notes**
```release-note
None
```

### CODE REVIEW

**Comment by nikopen** [comment]
@kubernetes/sig-apps-pr-reviews ping!

**Comment by runyontr** [comment]
Sorry for the delay on this. 

1. I think this solution goes agains the caching mechanism that's supposed to be used.  You may want me to rework this to do it better. Some alternative solutions
  a. Pass UIDs in the queue.
  b. Pass the correct data in the claimPods function to be handled, since that already fetches a fresh instance of the RS.  
2) This doesn't seem unique to ReplicaSets, but the race condition may occur more in this controller than others based on how Replicates are created/destroyed.  It may also not matter as much to other objects if the cached object isn't the same as the one on the server.
3) Since this issue occurred between the local informer cache and the remote server, it didn't seem possible to make a go test that would explicitly create the scenario required to simulate the race.    I created this PR https://github.com/kubernetes/kubernetes/pull/74452 To be able to run the `rs` test repeatedly looking for this race.  If someone has a better idea about how to put this into a test, I would like to implemented it for validation.

**Comment by nikopen** [comment]
/lgtm
/milestone v1.15

@kubernetes/sig-apps-pr-reviews ping!

File: pkg/controller/replicaset/replica_set.go [hunk 1]
```diff
@@ -580,6 +580,15 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
 		return err
 	}
 
+	//Do a full fetch here.
```
Comment (lavalamp) [inline pkg/controller/replicaset/replica_set.go]: You might as well replace the call to the lister.

nit: space between // and comment.

Comment (lavalamp) [reply pkg/controller/replicaset/replica_set.go]: Wait, I see this is detecting a deletion, so both calls are necessary.

File: pkg/controller/replicaset/replica_set.go [hunk 2]
```diff
@@ -580,6 +580,15 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
...
+	//The cache may be out of date via race condition #69376
+	fresh, err := rsc.kubeClient.AppsV1().ReplicaSets(rs.Namespace).Get(rs.Name, metav1.GetOptions{})
+	if err == nil && fresh.UID != rs.UID {
+		klog.Errorf("%v %v's UID in cache differed from server. This object was deleted.", rsc.Kind, key)
+		rsc.expectations.DeleteExpectations(key)
+		return nil
```
Comment (lavalamp) [inline pkg/controller/replicaset/replica_set.go]: Should this really return? Doesn't it need to proceed to process the new RS?

File: pkg/controller/replicaset/replica_set_test.go [hunk 3]
```diff
@@ -274,8 +275,10 @@ func TestSyncReplicaSetCreateFailures(t *testing.T) {
 	}
 }
 
-func TestSyncReplicaSetDormancy(t *testing.T) {
-	// Setup a test server so we can lie about the current state of pods
+func TestNonMatchingUIDs(t *testing.T) {
+	// Setup a test server that doesn't store the replica sets,
+	// So the returned Replica set wont have the UID and shouldn't
+	//process the event
```
Comment (lavalamp) [inline pkg/controller/replicaset/replica_set_test.go]: nit: missing space

File: pkg/controller/replicaset/replica_set.go (hunk 1 above)
Comment (lavalamp) [reply pkg/controller/replicaset/replica_set.go]: The opposite race can also happen, where the cache says "not found" even though it's just been added. I think the only consequence to that at the moment is a delay in processing until something else happens to the RS object.

File: pkg/controller/replicaset/replica_set.go (hunk 2 above)
Comment (runyontr) [reply pkg/controller/replicaset/replica_set.go]: The creation of the new RS shows up as a second event, and will get processed correctly there.  The event being handled here is the deletion of the first replicaset

Comment (lavalamp) [reply pkg/controller/replicaset/replica_set.go]: Triggers get merged by the work queue, so I don't know that that's guaranteed to be the case.

File: pkg/controller/replicaset/replica_set.go [hunk 4]
```diff
@@ -635,9 +644,6 @@ func (rsc *ReplicaSetController) claimPods(rs *apps.ReplicaSet, selector labels.
 		if err != nil {
 			return nil, err
 		}
-		if fresh.UID != rs.UID {
```
Comment (lavalamp) [inline pkg/controller/replicaset/replica_set.go]: why is this removed? it seems important?

Comment (runyontr) [reply pkg/controller/replicaset/replica_set.go]: Good catch.  At one point I was checking UIDs elsewhere and backed out those changes.  This looks like it was errantly removed as part of that cleanup.

File: test/cmd/apps.sh [hunk 5]
```diff
@@ -517,7 +517,7 @@ run_rs_tests() {
   # Pre-condition: no replica set exists
   kube::test::get_object_assert rs "{{range.items}}{{${id_field:?}}}:{{end}}" ''
   # Command
-  kubectl create -f hack/testdata/frontend-replicaset.yaml "${kube_flags[@]:?}"
```
Comment (lavalamp) [inline test/cmd/apps.sh]: Is this change necessary?

File: pkg/controller/replicaset/replica_set.go (hunk 2 above)
Comment (lavalamp) [reply pkg/controller/replicaset/replica_set.go]: still not convinced this early exit is a good idea.

**Comment by soggiest** [comment]
Hello! Code Freeze is just about on us. We'll be entering into Freeze tomorrow, May 31st. Is this still planned to land in 1.15?

**Comment by xmudrii** [comment]
@runyontr Hello! I'm bug triage lead for the 1.16 release cycle and considering this PR has not been updated for a long time, I'd like to check what's the status of this PR. The code freeze is starting 29th August (about 2.5 weeks from now) and while there is plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.16, is it still planned for this release?

**Comment by xmudrii** [comment]
@lavalamp @runyontr @janetkuo I've also checked on the issue, but this PR are open for a long time. Should we move this to the next milestone or should we remove this from the milestone entirely?

File: pkg/controller/replicaset/replica_set.go (hunk 2 above)
Comment (liggitt) [reply pkg/controller/replicaset/replica_set.go]: I don't think a live lookup is the correct fix here. It would make more sense to me to track the replicaset uid alongside the associated expectations, and if it mismatches, delete the expectations on sync

**Comment by liggitt** [comment]
no, this is not ready for merge.

/milestone v1.17

**Comment by josiahbjorgaard** [comment]
Bug triage for 1.17 here. Code freeze is Nov. 18th. Will this PR be merged by then? If not, can we move it out of the milestone?

**Comment by liggitt** [comment]
/close

in favor of https://github.com/kubernetes/kubernetes/pull/82572
